<a href="https://colab.research.google.com/github/shreya2118/Deep_learning/blob/main/CustomerChurnPrediction_backpropagation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

BINARY CLASSIFICATION PROBLEM--

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [5]:
df=pd.read_csv("Churn_Modelling.csv")
df.head()
#df.shape

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


In [7]:
df.duplicated().sum()

np.int64(0)

In [8]:
df['Exited'].value_counts()

Exited
0    7963
1    2037
Name: count, dtype: int64

In [9]:
df['Geography'].value_counts()

Geography
France     5014
Germany    2509
Spain      2477
Name: count, dtype: int64

Removing first three columns

we have used "inplace" to make permanent changes

In [16]:
df.drop(columns=['RowNumber','CustomerId','Surname'],inplace=True)

KeyError: "['RowNumber', 'CustomerId', 'Surname'] not found in axis"

In [15]:
df.head()

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619    France  Female   42       2       0.00              1   
1          608     Spain  Female   41       1   83807.86              1   
2          502    France  Female   42       8  159660.80              3   
3          699    France  Female   39       1       0.00              2   
4          850     Spain  Female   43       2  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0          1               1        101348.88       1  
1          0               1        112542.58       0  
2          1               0        113931.57       1  
3          0               0         93826.63       0  
4          1               1         79084.10       0

applying one-hot encoding to categorial columns

In [22]:
df=pd.get_dummies(df,columns=['Geography','Gender'],drop_first=True)

KeyError: "None of [Index(['Geography', 'Gender'], dtype='object')] are in the [columns]"

In [23]:
df

CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0             619   42       2       0.00              1          1   
1             608   41       1   83807.86              1          0   
2             502   42       8  159660.80              3          1   
3             699   39       1       0.00              2          0   
4             850   43       2  125510.82              1          1   
...           ...  ...     ...        ...            ...        ...   
9995          771   39       5       0.00              2          1   
9996          516   35      10   57369.61              1          1   
9997          709   36       7       0.00              1          0   
9998          772   42       3   75075.31              2          1   
9999          792   28       4  130142.79              1          1   

      IsActiveMember  EstimatedSalary  Exited  Geography_Germany  \
0                  1        101348.88       1              False   
1                  1        112542.58       0              False   
2                  0        113931.57       1              False   
3                  0         93826.63       0              False   
4                  1         79084.10       0              False   
...              ...              ...     ...                ...   
9995               0         96270.64       0              False   
9996               1        101699.77       0              False   
9997               1         42085.58       1              False   
9998               0         92888.52       1               True   
9999               0         38190.78       0              False   

      Geography_Spain  Gender_Male  
0               False        False  
1                True        False  
2               False        False  
3               False        False  
4                True        False  
...               ...          ...  
9995            False         True  
9996            False         True  
9997            False        False  
9998            False         True  
9999            False        False  

[10000 rows x 12 columns]

For scaling, first perform train test split, and then standardize

In [24]:
X=df.drop(columns=['Exited'])
y=df['Exited']
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=1)

In [28]:
X_train.shape

(8000, 11)

In [29]:
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()

X_train_scaled=scaler.fit_transform(X_train)
X_test_scaled=scaler.transform(X_test)

In [30]:
X_train_scaled

array([[-0.23082038, -0.94449979, -0.70174202, ...,  1.71490137,
        -0.57273139,  0.91509065],
       [-0.25150912, -0.94449979, -0.35520275, ..., -0.58312392,
        -0.57273139, -1.09278791],
       [-0.3963303 ,  0.77498705,  0.33787579, ...,  1.71490137,
        -0.57273139, -1.09278791],
       ...,
       [ 0.22433188,  0.58393295,  1.3774936 , ..., -0.58312392,
        -0.57273139, -1.09278791],
       [ 0.13123255,  0.01077067,  1.03095433, ..., -0.58312392,
        -0.57273139, -1.09278791],
       [ 1.1656695 ,  0.29735181,  0.33787579, ...,  1.71490137,
        -0.57273139,  0.91509065]])

In [31]:
import tensorflow as tf
from tensorflow import keras
from keras import Sequential
from keras.layers import Dense

Sequential model

By this we added our input and output layer

In [32]:
model=Sequential()
#1 i/p layer, 1 hidden layer= 3 perceptrons/nodes, 1 o/p layer

model.add(Dense(3,activation='sigmoid',input_dim=11))
model.add(Dense(1,activation='sigmoid'))


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [33]:
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 3)              │            36 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 1)              │             4 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 40 (160.00 B)

 Trainable params: 40 (160.00 B)

 Non-trainable params: 0 (0.00 B)

In [34]:
model.compile(loss='binary_crossentropy',optimizer='Adam')

#optimizer is used for gradient descent

In [36]:
model.fit(X_train_scaled,y_train,epochs=10)
#model gets trained

Epoch 1/10
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - loss: 0.4277
Epoch 2/10
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 0.4236
Epoch 3/10
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 0.4158
Epoch 4/10
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 0.4242
Epoch 5/10
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 0.4233
Epoch 6/10
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 0.4242
Epoch 7/10
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 0.4126
Epoch 8/10
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 0.4274
Epoch 9/10
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 0.4082
Epoch 10/10
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 0.4173


In [43]:
#weights ki value and bias=3 ki value
#input layer or next hidden layer ke weights h

model.layers[0].get_weights()

[array([[ 0.09229597, -0.06553625,  0.0577851 ],
        [-0.635643  ,  0.41008157, -2.7144327 ],
        [-0.18534374, -0.38224432,  0.03895672],
        [-0.6526665 ,  0.575824  ,  0.14570117],
        [ 0.15737557,  0.10201245,  0.00895533],
        [ 0.12401045,  0.06130512, -0.03142039],
        [ 0.8611216 , -1.5759714 ,  0.128923  ],
        [-0.11053571,  0.25694674,  0.12696709],
        [-0.46409273,  0.7874356 , -0.3996219 ],
        [-0.12267853, -0.07136784, -0.01993127],
        [ 0.69235337, -0.46991387,  0.15035737]], dtype=float32),
 array([ 0.2808609 , -0.6199969 ,  0.45047408], dtype=float32)]

In [44]:
#hidden layer or output layer ke beech ke weights and bias

model.layers[1].get_weights()

[array([[-1.2011766 ],
        [ 0.85026926],
        [-1.8807355 ]], dtype=float32),
 array([-0.22976488], dtype=float32)]

Now, the model is fully trained, and we can use it now for the prediction

In [46]:
z=model.predict(X_test_scaled)

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step  


The values are not in 0/1 because we have used sigmoid function, we can change it to 0/1 by setting a **threshold** and then saying if it is greater than the threshold, output that as 1 or 0

In [47]:
y_pred=np.where(z>0.5,1,0)

In [48]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test,y_pred)

0.8235

**Right now, the accuracy is 0.8235, we can increase the accuracy by **

*  increasing Epochs
*  activation function set to "ReLu"
*  increase the number of nodes of hidden layers
*  increase the number of hidden layers




**CHANGED CODE**

In [50]:
model=Sequential()
#1 i/p layer, 2 hidden layer= 11 perceptrons/nodes,, 1 o/p layer

model.add(Dense(11,activation='relu',input_dim=11))
model.add(Dense(11,activation='relu'))
model.add(Dense(1,activation='sigmoid'))


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [51]:
model.summary()

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_2 (Dense)                 │ (None, 11)             │           132 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 11)             │           132 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 1)              │            12 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 276 (1.08 KB)

 Trainable params: 276 (1.08 KB)

 Non-trainable params: 0 (0.00 B)

In [52]:
model.compile(loss='binary_crossentropy',optimizer='Adam',metrics=['accuracy'])

KEEP IN MIND, THAT TRAING ACCURACY AS WELL AS VALIDATION ACCURACY BOTH SHOULD INCREASE,

---

IF VALIDATION ACCURACY IS NOT INCREASING, IT BECOMES THE CASE OF OVERFITTING

In [54]:
model.fit(X_train_scaled,y_train,epochs=25,validation_split=0.2)


Epoch 1/25
200/200 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8645 - loss: 0.3264 - val_accuracy: 0.8506 - val_loss: 0.3551
Epoch 2/25
200/200 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8612 - loss: 0.3379 - val_accuracy: 0.8525 - val_loss: 0.3524
Epoch 3/25
200/200 ━━━━━━━━━━━━━━━━━━━━ 3s 12ms/step - accuracy: 0.8701 - loss: 0.3253 - val_accuracy: 0.8487 - val_loss: 0.3524
Epoch 4/25
200/200 ━━━━━━━━━━━━━━━━━━━━ 1s 6ms/step - accuracy: 0.8658 - loss: 0.3208 - val_accuracy: 0.8500 - val_loss: 0.3525
Epoch 5/25
200/200 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.8635 - loss: 0.3309 - val_accuracy: 0.8531 - val_loss: 0.3541
Epoch 6/25
200/200 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8629 - loss: 0.3318 - val_accuracy: 0.8512 - val_loss: 0.3531
Epoch 7/25
200/200 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8609 - loss: 0.3421 - val_accuracy: 0.8519 - val_loss: 0.3528
Epoch 8/25
200/200 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8699 - loss: 0.3194 - val_accuracy: 0

In [56]:
model.layers[2].get_weights()

[array([[-0.8617016 ],
        [ 0.63893443],
        [-0.8941313 ],
        [ 0.8445291 ],
        [-0.52506536],
        [ 0.8479343 ],
        [ 0.5957352 ],
        [-0.79547733],
        [ 1.348493  ],
        [-0.9544252 ],
        [-0.5939206 ]], dtype=float32),
 array([-0.21610002], dtype=float32)]

In [57]:
z=model.predict(X_test_scaled)

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step


In [58]:
y_pred=np.where(z>0.5,1,0)

In [59]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test,y_pred)

0.862

**ACCURACY HAS BEEN IMPROVED TO** 0.862